src/dmd/dmodule.d

```diff
@@ -653,14 +653,127 @@ extern (C++) final class Module : Package
+            {
+                uint u2;
+                buf = buf[1..$];
+                if (!buf.length)
+                {
+                    error("surrogate UTF-16 high value %04x at end of file", u);
+                    fatal();
+                }
+                u2 = readNext(buf.ptr);
+                if (u2 < 0xDC00 || u2 > 0xDFFF)
+                {
+                    error("surrogate UTF-16 low value %04x out of range", u2);
+                    fatal();
+                }
+                u = (u - 0xD7C0) << 10;
+                u |= (u2 - 0xDC00);
+            }
+            else if (u >= 0xDC00 && u <= 0xDFFF)
+            {
+                error("unpaired surrogate UTF-16 value %04x", u);
+                fatal();
+            }
+            else if (u == 0xFFFE || u == 0xFFFF)
+            {
+                error("illegal UTF-16 value %04x", u);
+                fatal();
+            }
+            return u;
+
+        }
+
+        /*
+         * Convert a buffer from UTF16/32 to UTF8
+         * Params:
+         *    SourceEncoding    is the source UTF 16 or 32?
+         *    Endian            is the buffer big/little endian
+         *    buf               buffer of unicode data
+         * Returns:
+         *    input buffer reencoded as UTF8
+         */
+        char[] toUTF8(SourceEncoding sourceEncoding, Endian endian)(const(char)[] buf)
+        {
+            static if (sourceEncoding == SourceEncoding.utf16)
+            {
+                alias SourceElement = ushort;
+                static if (endian == Endian.little) alias readNext = Port.readwordLE;
+                else alias readNext = Port.readwordBE;
+            }
+            else
+            {
+                alias SourceElement = uint;
+                static if ( endian == Endian.little) alias readNext = Port.readlongLE;
+                else alias readNext = Port.readlongBE;
+            }
+            if (buf.length & (SourceElement.sizeof -1))
+            {
+                error("odd length of UTF-%d char source %u", SourceElement.sizeof*8, buf.length);
+                fatal();
+            }
+
+            const(SourceElement)[] eBuf = cast(const(SourceElement)[])buf;
+
+            OutBuffer dbuf;
+            dbuf.reserve(eBuf.length);
+
+            for(; eBuf.length; eBuf = eBuf[1..$])
+            {
+                uint u = readNext(eBuf.ptr);
+                if (u & ~0x7F)
+                {
+                    static if (sourceEncoding == SourceEncoding.utf32)
+                    {
+                        if (u > 0x10FFFF)
+                        {
+                            error("UTF-32 value %08x greater than 0x10FFFF", u);
+                            fatal();
+                        }
+                        dbuf.writeUTF8(u);
+                    }
+                    else //utf16
+                    {
+                        dbuf.writeUTF8(finishNonASCIICodepoint!readNext(u, eBuf));
```

Comment:
In fact, it's weirdly already two functions, as finishNonASCIICodepoint is only for UTF16.

Reply:
It seems to me that the UTF16/32 versions do basically the same thing and the only implementation differences are the source element type, and how to deal with non ascii code points, but maybe I'm being overly general.  

I'll split them up into separate functions for UTF32 and UTF16 inputs